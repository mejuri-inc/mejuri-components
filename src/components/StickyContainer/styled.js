import styled from 'styled-components'

export const Wrapper = styled.div`
  min-height: ${(p) => `${p.height}px`};
  width: 100%;
`

export const Content = styled.div`
  position: ${(p) => (p.sticked ? 'fixed' : 'relative')};
  top: 0;
  width: 100%;
`
