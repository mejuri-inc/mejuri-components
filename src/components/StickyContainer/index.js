import React from 'react'
import PropTypes from 'prop-types'
import { Content, Wrapper } from './styled'

class StickyContainer extends React.Component {
  constructor(props) {
    super(props)
    this.r = React.createRef()

    this.state = {
      sticked: false,
      stickPosition: null,
      height: null
    }

    this.scroll = this.scroll.bind(this)
  }

  scroll(e) {
    const { sticked, stickPosition } = this.state
    if (window.scrollY >= stickPosition && !sticked) {
      this.setState({ sticked: true })
    }
    if (window.scrollY < stickPosition && sticked) {
      this.setState({ sticked: false })
    }
  }

  componentDidMount() {
    const rect = this.r.current.getBoundingClientRect()
    this.setState(
      {
        sticked: window.scrollY >= this.r.current.offsetTop,
        stickPosition: this.r.current.offsetTop,
        height: rect.height
      },
      window && window.addEventListener('scroll', this.scroll)
    )
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.scroll)
  }

  render() {
    const { height, sticked } = this.state
    return (
      <Wrapper ref={this.r} height={height}>
        <Content sticked={sticked}>{this.props.children}</Content>
      </Wrapper>
    )
  }
}

StickyContainer.propTypes = {
  children: PropTypes.node
}

export default StickyContainer
