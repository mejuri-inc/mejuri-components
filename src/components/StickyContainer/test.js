import React from 'react'
import { render } from '@testing-library/react'
import 'jest-styled-components'
import StickyContainer from './index'

describe('StickyContainer', () => {
  it('renders ok', () => {
    const { container } = render(<StickyContainer />)
    expect(container).toMatchSnapshot()
  })

  it('renders children', () => {
    const SomeComponent = () => <div data-test-id='i-am-the-child' />

    const { queryByTestId } = render(
      <StickyContainer>
        <SomeComponent />
      </StickyContainer>
    )

    expect(queryByTestId('i-am-the-child')).toBeDefined()
  })
})
