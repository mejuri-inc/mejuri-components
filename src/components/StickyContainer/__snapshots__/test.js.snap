// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickyContainer renders ok 1`] = `
.c0 {
  min-height: 0px;
  width: 100%;
}

.c1 {
  position: relative;
  top: 0;
  -webkit-transition: top 0.5s ease;
  transition: top 0.5s ease;
  width: 100%;
  z-index: 11;
}

<div>
  <div
    class="c0"
    height="0"
  >
    <div
      class="c1"
      data-h="header-visible"
    />
  </div>
</div>
`;
